### PR TITLE
persist-txn: hook enable_persist_txn_tables up to Launch Darkly

### PR DESCRIPTION
**Pull request by danhhz**
Only takes effect on restart. Any changes will also cause clusterd processes to restart.

The usual Launch Darkly system vars hookup is not available early enough in the boot process to use directly, so we mirror it to an entry in the catalog stash "config" collection and read that. Leave the environmentd cli flag around as an override to make testing easier.

Touches https://github.com/MaterializeInc/materialize/issues/22173
Touches https://github.com/MaterializeInc/materialize/issues/22801

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/ddl.rs [hunk 1]
```diff
@@ -525,6 +529,9 @@ impl Coordinator {
             if update_http_config {
                 self.update_http_config();
             }
+            if update_enable_persist_txn_tables {
+                self.update_txns_config().await;
+            }
```
Comment (jkosh44) [inline src/adapter/src/coord/ddl.rs:534]: I think you're betting off removing this and updating the following method: https://github.com/MaterializeInc/materialize/blob/5b3586ffa4396e08be157bb984d22af07f5b40c5/src/adapter/src/catalog.rs#L3195-L3205

You would have to add a method to `Transaction`, if one doesn't already exist, then something the following to the end:
```Rust
        if name == ENABLE_PERSIST_TXN_TABLES {
            tx.set_enable_persist_txn_tables(state.system_configuration.enable_persist_txn_tables());
        }
```

That way both the config and the system var values will be updated in the same transaction. With the current approach it seems like the two values could end up drifting apart if we get a failure after the catalog transaction commits but before this is run.

Also you'd be able to remove `set_enable_persist_txn_tables` from `DurableCatalogState`.

File: src/environmentd/src/lib.rs [hunk 2]
```diff
@@ -447,12 +453,15 @@ impl Listeners {
...
-        let openable_adapter_storage = catalog_opener(
+        let mut openable_adapter_storage = catalog_opener(
             &config.catalog_config,
             &config.controller,
             &config.environment_id,
         )
         .await?;
+        let enable_persist_txn_tables_stash_ld = openable_adapter_storage
+            .get_enable_persist_txn_tables()
+            .await?;
```
Comment (jkosh44) [inline src/environmentd/src/lib.rs:464]: Can you add a comment here to indicate that this isn't actually the current value of LD, but what the value in LD was the last time this environment was running?

File: src/sql/src/session/vars.rs [hunk 3]
```diff
@@ -1191,6 +1191,24 @@ pub static ENABLE_LAUNCHDARKLY: ServerVar<bool> = ServerVar {
...
+        one.
+
+        Only takes effect on restart. Any changes will also cause clusterd \
+        processes to restart.
+
+        This value is also configurable via a Launch Darkly parameter of the \
+        same name, but we keep the flag to make testing easier. If specified, \
+        the flag takes precedence over the Launch Darkly param.",
+    internal: true,
+};
```
Comment (jkosh44) [inline src/sql/src/session/vars.rs:1210]: We have the `feature_flags!` macro that help with defining features flags if you prefer to use that. You can see examples by grepping for `feature_flags!` in this file.

File: src/sqllogictest/src/runner.rs [hunk 4]
```diff
@@ -1043,6 +1041,8 @@ impl<'a> RunnerInner<'a> {
             deploy_generation: None,
             http_host_name: Some(host_name),
             internal_console_redirect_url: None,
+            // TODO(txn): Get this flipped to true before turning anything on in prod.
```
Comment (def-) [inline src/sqllogictest/src/runner.rs:1044]: This part is still key. Do we expect everything interesting to be covered by existing tests? I didn't follow this feature closely before.

**Review by def- — APPROVED** [review]

File: src/adapter/src/coord/ddl.rs (hunk 1 above)
Comment (danhhz) [reply src/adapter/src/coord/ddl.rs:534]: Yeah, that's much better, thanks!

File: src/environmentd/src/lib.rs (hunk 2 above)
Comment (danhhz) [reply src/environmentd/src/lib.rs:464]: Done

File: src/sqllogictest/src/runner.rs (hunk 4 above)
Comment (danhhz) [reply src/sqllogictest/src/runner.rs:1044]: Yeah, I promise I'll remember this! The `TODO(txn)` marker is one I'm grepping for and making sure they're all resolved before we turn it on in prod.

We'll definitely want some new testing, but it's not quite ready for it yet. Hoping within the next week or so! I've promised Philip that I'll ping QA when it's ready, feel free to continue ignoring this until then

File: src/sql/src/session/vars.rs (hunk 3 above)
Comment (danhhz) [reply src/sql/src/session/vars.rs:1210]: Done

**Review by danhhz — COMMENTED** [review]
The new `set_` approach was a big enough diff that I figured it would be easier to force push instead of an append-only fixup commit

**Comment by danhhz** [comment]
@jkosh44 Friendly ping on this! :)

**Comment by jkosh44** [comment]
> @jkosh44 Friendly ping on this! :)

Oops, I definitely lost track of this PR. I'll take a look now. If ever go quiet on a PR after starting a review for more than 24 hours without reaching out then feel free to shoot me a ping.

**Comment by danhhz** [comment]
No worries and will do! 👍

**Review by jkosh44 — APPROVED** [review]
Nice!

**Comment by danhhz** [comment]
TFTR!

**Pull request merged**